File: apps/replay/modules/f1-haptics/src/F1HapticsModule.ts
import { NativeModule, requireOptionalNativeModule } from 'expo';

import type {
  HapticTextureDefinition,
  TelemetryHapticUpdate,
} from './F1Haptics.types';

declare class F1HapticsModule extends NativeModule {
  isSupported(): boolean;

  playPattern(pattern: HapticTextureDefinition): Promise<void>;

  startTelemetry(): Promise<void>;

  stop(): Promise<void>;

  stopTelemetry(): Promise<void>;

  updateTelemetry(update: TelemetryHapticUpdate): Promise<void>;
}

export const F1Haptics =
  requireOptionalNativeModule<F1HapticsModule>('F1Haptics');

File: apps/replay/modules/f1-haptics/android/src/main/java/expo/modules/f1haptics/F1HapticsModule.kt
package expo.modules.f1haptics

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class F1HapticsModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("F1Haptics")

    Function("isSupported") {
      // TODO: Report Android waveform/composition support when that adapter is designed.
      false
    }

    AsyncFunction("playPattern") { _: Map<String, Any?> ->
      // TODO: Translate the shared texture definition to VibrationEffect primitives.
    }

    AsyncFunction("startTelemetry") {
      // Android intentionally remains a no-op until its waveform mapping is designed.
    }

    AsyncFunction("updateTelemetry") { _: Map<String, Any?> ->
      // Android intentionally remains a no-op until its waveform mapping is designed.
    }

    AsyncFunction("stopTelemetry") {
      // Android intentionally remains a no-op until its waveform mapping is designed.
    }

    AsyncFunction("stop") {
      // TODO: Cancel the active vibrator effect.
    }
  }
}
